pkg/alertmanager/multitenant.go

```diff
@@ -1225,3 +1191,30 @@ func storeTemplateFile(templateFilepath, content string) (bool, error) {
 
 	return true, nil
 }
+
+func configChanged(left, right alertspb.AlertConfigDesc) bool {
+	if left.User != right.User {
+		return true
+	}
+	if left.RawConfig != right.RawConfig {
+		return true
+	}
+
+	existing := make(map[string]string)
+	for _, tm := range left.Templates {
+		existing[tm.Filename] = tm.Body
+	}
+
+	for _, tm := range right.Templates {
+		corr, ok := existing[tm.Filename]
+		if !ok {
+			return true // Right has a template that left does not.
+		}
+		if corr != tm.Body {
+			return true // The template content is different.
+		}
+		delete(existing, tm.Filename)
+	}
+
+	return len(existing) != 0 // Left has a template that right does not.
```

Comment:
[nit] I'd compare the length first, instead of deleting and comparing to zero, then the function ends with `return false`. This is probably subjective - I'm just more used to seeing it done that way.

```suggestion
        if len(existing) != len(right.Templates) {
                return true
        }

	for _, tm := range right.Templates {
		corr, ok := existing[tm.Filename]
		if !ok {
			return true // Right has a template that left does not.
		}
		if corr != tm.Body {
			return true // The template content is different.
		}
	}

	return false
```

Reply:
Actually this might not work if there are templates with the same `Filename` in `right.Templates` - feel free to ignore this.

Reply:
> if there are templates with the same Filename

This is exactly the rationale here. I wasn't 100% certain that validation would have run before this code to ensure this in all paths (and, it felt cleaner to not tie the code to that assumption).

Would you feel that putting the right side into a map would be cleaner in terms of readability? It requires some extra allocations and passes, but I believe the limit for template count is usually pretty low.